packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/ArgumentProvider.cs

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.Generator.CSharp.Expressions;
+using static Microsoft.Generator.CSharp.Expressions.Snippets;
+
+namespace Microsoft.Generator.CSharp
+{
+    internal class ArgumentProvider : TypeProvider
+    {
+        private static readonly Lazy<ArgumentProvider> _instance = new(() => new ArgumentProvider());
+
+        private class Template<T> { }
+
+        private const string AssertNotNullMethodName = "AssertNotNull";
+        private const string AssertNotNullOrEmptyMethodName = "AssertNotNullOrEmpty";
+        private const string AssertNotNullOrWhiteSpaceMethodName = "AssertNotNullOrWhiteSpace";
+
+        private readonly CSharpType _t = typeof(Template<>).GetGenericArguments()[0];
+        private readonly Parameter _nameParam = new Parameter("name", null, typeof(string), null, ValidationType.None, null);
+        private readonly CSharpType _nullableT;
+        private readonly ParameterReference _nameParamRef;
+
+        public static ArgumentProvider Instance => _instance.Value;
+
+        private ArgumentProvider() : base(null)
+        {
+            _nameParamRef = new ParameterReference(_nameParam);
+            _nullableT = _t.WithNullable(true);
```

Comment:
Currently, `_nullableT` returns `T` instead of `T?`. Tracking issue here: https://github.com/microsoft/typespec/issues/3490